4_evaluate/src/interpolate_predictions.py

```diff
@@ -0,0 +1,225 @@
+    :param sequence_indices: Index of every observation's corresponding sequence in predictions array
+    :param date_indices: Index for every observation within its sequence
+    :param depth_above_indices: Indexing array for the predictions above the observation locations
+    :param depth_below_indices: Indexing array for the predictions below the observation locations
+    :param above_factors: Interpolation factors for the predictions above the observations
+    :param below_factors: Interpolation factors for the predictions below the observations
+    :param use_obs: Indexing array for observation locations to interpolate to
+    :returns: Array of temperature predictions interpolated to observation depths
+
+    """
+    # Load predictions
+    predictions_npz = np.load(predictions_filepath)
+    ta = predictions_npz['predictions'][sequence_indices[use_obs], date_indices[use_obs], depth_above_indices[use_obs]]
+    tb = predictions_npz['predictions'][sequence_indices[use_obs], date_indices[use_obs], depth_below_indices[use_obs]]
+    # Interpolate
+    predictions = ta*above_factors[use_obs] + tb*below_factors[use_obs]
+    return predictions
+
+
+def get_interpolated_predictions(predictions_filepath,
+                                 metadata_filepath,
+                                 dataset_filepath,
+                                 obs_file,
+                                 interpolated_predictions_filepath,
+                                 batch_size=5000):
+    """
+    Interpolate ML output to observation depths.
+
+    :param predictions_filepath: Path to unnormalized predictions for one
+        dataset (train, validation, or test)
+    :param metadata_filepath: Path to model training metadata npz for one dataset
+    :param dataset_filepath: Path to input for one dataset
+    :param obs_file: File with all temperature observations
+    :param interpolated_predictions_filepath: csv file to save interpolated
+        predictions to
+    :param batch_size: Number of observations to use per batch while matching
+        observations to prediction sequences (Default value = 5000)
+
+    """
+
+    # Load observations... from original data!
+    # Load normalized observations to get site ids
+    dataset_npz = np.load(dataset_filepath)
+    site_ids = dataset_npz['site_ids']
+    start_dates = dataset_npz['start_dates']
+    sequence_length = dataset_npz['sequence_length']
+    spinup_time = dataset_npz['spinup_time']
+
+    # Load metadata to get depths_use
+    metadata_npz = np.load(metadata_filepath)
+    depths_use = metadata_npz['depths_use']
+
+    # Now, we'll need obs for each of those lakes.
+    # Read temperature observations
+    obs_all = pd.read_csv(obs_file, index_col=0, parse_dates=['date'])
+    # Whittle down to lakes in this dataset
+    obs = obs_all[obs_all.site_id.isin(site_ids)]
+    obs_all.site_id.unique().shape, obs.site_id.unique().shape, np.unique(site_ids).shape
```

Comment:
Does this actually print from snakemake / is it useful or is this a development relic?

If it is a development aid, perhaps an `assert(shape1 == shape2)` (like those you have later) would be a useful test to have here instead.

Reply:
Whoops! Yes, this is a pure dev relic. I think I'll leave out asserts for now